sdk/avalon_sdk/interfaces/event_processor.py

```diff
@@ -0,0 +1,186 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import logging
+from abc import ABC, abstractmethod
+
+logging.basicConfig(
+    format='%(asctime)s - %(levelname)s - %(message)s', level=logging.INFO)
+
+LISTENER_SLEEP_DURATION = 2  # second
+
+class EventProcessor(ABC):
+
+    async def listener(self, event_filter, *kargs, **kwargs):
+        """
+        This function defines a listener that runs an infinite loop
+        to listen for events from an event source. It might use pass
+        on the event_filter to the blockchain specific get_events()
+        fucntion call. It puts the raw events into a shared queue
+        which handlers would be reading from. It sleeps for a short
+        duration of LISTENER_SLEEP_DURATION seconds after fetching
+        and putting into the queue a batch of events.
+        """
+        logging.info("Started listener for events from blockchain")
+        while True:
+            # Check for any new event logs since the last poll
+            # on this filter. Though using events, this is not
+            # fully asynchronous.
+            for event in self.get_events(event_filter, *kargs, **kwargs):
+                await self.queue.put(event)
+                logging.debug("Event pushed into listener Queue")
+            await asyncio.sleep(LISTENER_SLEEP_DURATION)
+
+    @abstractmethod
+    def get_events(self, event_filter, *kargs, **kwargs):
+        """
+        This function is meant to be implemented by blockchain specific
+        event processor. It gets events from the event source on the
+        basis of the event_filter passed in.
+        Returns:
+            An iterable set/list of events received from event source.
+        """
+        pass
+
+    async def handler(self, callback, *kargs, **kwargs):
+        """
+        This function defines a typical event handler that waits on
+        a shared queue to get events in an infinite loop. Once it gets
+        an event it passes it on to process_and_delegate() which
+        thereafter might process and delegate to the callback function.
+        """
+        logging.info("Started handler to handle events")
+        while True:
+            event = await self.queue.get()
+            logging.debug("Event popped from listener queue")
+            self.process_and_delegate(event, callback, *kargs, **kwargs)
+            self.queue.task_done()
+
+    @abstractmethod
+    def process_and_delegate(self, event, callback, *kargs, **kwargs):
+        """
+        This function is meant to be implemented by blockchain specific
+        event processor. It processes the event passed in to it and then
+        passes on the processed event to the callback function for
+        appropriate handling.
```

Comment:
Not sure of the use of callback in process_and_delegate(). How to process the events is already implementations specific.